### PR TITLE
move drive pagers to pager pattern

### PR DESCRIPTION
**Pull request by ryanfkeepers**
Drive pager usage currently showcases strong coupling between two layers: drive collection logic processing and drive api.  This PR separates that coupling by moving the full item enumeration process into the
API, and letting the collection logic process the
results. This acs as both a simplification of complex code, and a clearer separation of ownership between the two layers.

A detrimental side effect of this change is that drive item enumeration has moved from page-streaming
(ie: each page is fully processed before moving on to the next) and onto batch processing (ie: all items are stored in memory and processed in a single pass). Acknowledging that this is an unacceptable regression, a follow-up PR will appear shortly with better handling for stream-processing enumeration from the API layer as a standard part of the pattern for all pager
implementations.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :broom: Tech Debt/Cleanup

#### Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Comment by ashmrtn** [comment]
this seems like it could use a rebase prior to review

File: src/internal/m365/service/onedrive/mock/handlers.go [hunk 1]
```diff
@@ -22,6 +24,8 @@ import (
 type BackupHandler struct {
 	ItemInfo details.ItemInfo
 
+	EDID EnumeratesDriveItemsDelta
```
Comment (ashmrtn) [inline src/internal/m365/service/onedrive/mock/handlers.go:27]: please use more meaningful names for variables. EDID gives no context as to what the variable actually holds

File: src/internal/m365/collection/drive/url_cache_test.go [hunk 2]
```diff
@@ -509,63 +449,3 @@ func (suite *URLCacheUnitSuite) TestNeedsRefresh() {
 	cache.lastRefreshTime = time.Now()
 	require.False(t, cache.needsRefresh())
 }
-
-// Test newURLCache
-func (suite *URLCacheUnitSuite) TestNewURLCache() {
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/url_cache_test.go:514]: we should probably keep these tests. They check important edge cases that would be good to have "documented" somewhere so we know the behavior doesn't change

File: src/internal/m365/collection/drive/url_cache_test.go [hunk 3]
```diff
@@ -376,23 +330,10 @@ func (suite *URLCacheUnitSuite) TestGetItemProperties() {
 			},
 		},
 		{
-			name: "multi-page delta query error",
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/url_cache_test.go:379]: The original test here is important to keep around because it concerns possible partial results that we may receive. We can't assume that delta queries will always either complete successfully or fail on the first call

**Review by ashmrtn — CHANGES_REQUESTED** [review]

File: src/internal/m365/collection/drive/collections.go [hunk 4]
```diff
@@ -257,76 +257,78 @@ func (c *Collections) Get(
...
 
 	var (
-		// Drive ID -> delta URL for drive
-		deltaURLs = map[string]string{}
-		// Drive ID -> folder ID -> folder path
-		folderPaths  = map[string]map[string]string{}
-		numPrevItems = 0
+		driveIDToDeltaLink = map[string]string{}
+		driveIDToPrevPaths = map[string]map[string]string{}
+		numPrevItems       = 0
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:262]: Beats me.  Maybe @pandeyabs knows what it's for?

File: src/internal/m365/collection/drive/collections.go [hunk 5]
```diff
@@ -257,76 +257,78 @@ func (c *Collections) Get(
...
 		var (
-			driveID     = ptr.Val(d.GetId())
-			driveName   = ptr.Val(d.GetName())
-			prevDelta   = prevDeltas[driveID]
-			oldPaths    = oldPathsByDriveID[driveID]
-			numOldDelta = 0
-			ictx        = clues.Add(ctx, "drive_id", driveID, "drive_name", driveName)
+			driveID   = ptr.Val(d.GetId())
+			driveName = ptr.Val(d.GetName())
+			ictx      = clues.Add(ctx, "drive_id", driveID, "drive_name", driveName)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:269]: Happy to!

File: src/internal/m365/collection/drive/collections.go [hunk 6]
```diff
@@ -357,7 +359,7 @@ func (c *Collections) Get(
 				return nil, false, clues.Wrap(err, "making exclude prefix").WithClues(ictx)
 			}
 
-			ssmb.Add(p.String(), excluded)
+			ssmb.Add(p.String(), excludedItemIDs)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:362]: I'm happy to take recommendations here.

File: src/internal/m365/collection/drive/collections.go [hunk 7]
```diff
@@ -758,13 +769,13 @@ func (c *Collections) UpdateCollections(
...
 
-			prevPathStr, ok := oldPaths[itemID]
+			prevPathStr, ok := oldPrevPaths[itemID]
 			if ok {
 				prevPath, err = path.FromDataLayerPath(prevPathStr, false)
 				if err != nil {
 					el.AddRecoverable(ctx, clues.Wrap(err, "invalid previous path").
 						WithClues(ictx).
-						With("path_string", prevPathStr))
+						With("prev_path_string", prevPathStr))
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:778]: can do!

File: src/internal/m365/collection/drive/url_cache_test.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/url_cache_test.go:379]: That's no longer in scope of the call.  The enumerator, as of this change, either completely batches the items or returns an error.  It's not possible to have a partially complete query.

I probably need to re-add them to the next PR, however, since we re-add streaming pagination streaming.  I don't think I did that on the first pass.  Look forward to these re-appearing there, please.

File: src/internal/m365/collection/drive/url_cache_test.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/url_cache_test.go:514]: agreed and will do.  Can't remember why I originally took them out.

File: src/internal/m365/collection/drive/collections.go [hunk 8]
```diff
@@ -257,76 +257,78 @@ func (c *Collections) Get(
...
 		}
 
 		// Avoid the edge case where there's no paths but we do have a valid delta
 		// token. We can accomplish this by adding an empty paths map for this
 		// drive. If we don't have this then the next backup won't use the delta
 		// token because it thinks the folder paths weren't persisted.
-		folderPaths[driveID] = map[string]string{}
-		maps.Copy(folderPaths[driveID], paths)
+		driveIDToPrevPaths[driveID] = map[string]string{}
+		maps.Copy(driveIDToPrevPaths[driveID], newPrevPaths)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:326]: > The usage of this variable is extremely confusing and difficult to track. It's also not clear that it results in correct behavior under some edge cases

Oh, believe me, I am 100% there with you.  That said, these requests are a bit out of scope of this change.  The goal was only to refactor the drive enumeration, and to make the minimum amount of logic changes to the code calling it.  I think I've been able to do that, and am certainly open to the possibility that I messed something up because, like you said, this is complicated as all get-out.

As for a sanity refactor to this code itself, that's another task, and I've already completely overblown my timebox on this change to begin with.  I'll create a tech debt ticket for it, but would prefer to keep the work to that space.

[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:326]: issue: https://github.com/alcionai/corso/issues/4379

[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections.go:329]: Can we at least go back to not reusing the same variable for different purposes? IIRC, the original code used the per-drive item set for detecting deleted items. I'm fine with addressing some of this in a follow up refactor, but I'm not comfortable with making the code more difficult to read by reusing a variable like this

File: src/internal/m365/collection/drive/collections.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/m365/collection/drive/collections.go:365]: well, right now the variable name is describing the type of the thing, not the function. This variable is used to track the set of items to delete from kopia so something like `excludedItemsByDrive` could work

**Review by ashmrtn — APPROVED** [review]
approving to unblock, would like to see changes around the variable aliasing and a naming though

File: src/internal/m365/collection/drive/collections.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/collections.go:329]: sure, can do
